modules/core/src/main/java/org/apache/ignite/internal/processors/query/GridQueryProcessor.java

```diff
@@ -1996,9 +2017,13 @@ private boolean rebuildIsMeaningless(GridCacheContext cctx) {
      * Rebuilds indexes for provided caches from corresponding hash indexes.
      *
      * @param cctx Cache context.
+     * @param execSvc Thread pool for rebuilding indexes.
      * @return Future that will be completed when rebuilding is finished.
      */
-    public IgniteInternalFuture<?> rebuildIndexesFromHash(GridCacheContext cctx) {
+    public IgniteInternalFuture<?> rebuildIndexesFromHash(GridCacheContext cctx, ExecutorService execSvc) {
```

Comment:
What the point in having an argument for executor and exposing executor at all ?
Looks like you can use single instance allowing threads timeout and doesn't allowing more threads than number of cores.
I see no need in dynamic executor management.
The logic should be placed to org.apache.ignite.internal.GridKernalContext

exec = new IgniteThreadPoolExecutor(
            "rebuild-idx-runner",
            ctx.igniteInstanceName(),
            0,
            parallelism,
            DFLT_THREAD_KEEP_ALIVE_TIME,
            new LinkedBlockingQueue<>(),
            GridIoPolicy.UNDEFINED,
            oomeHnd
        );

exec .allowCoreThreadTimeOut(true);